src/Servers/Kestrel/Kestrel/test/KestrelConfigurationLoaderTests.cs

```diff
@@ -252,6 +252,112 @@ public void ConfigureEndpointDevelopmentCertificateGetsLoadedWhenPresent()
+            var certificate = new X509Certificate2(TestResources.GetCertPath("aspnetdevcert.pfx"), "testPassword", X509KeyStorageFlags.Exportable);
+            var bytes = certificate.Export(X509ContentType.Pkcs12, "1234");
+            var path = GetCertificatePath();
+            Directory.CreateDirectory(Path.GetDirectoryName(path));
+            File.WriteAllBytes(path, bytes);
+
+            var config = new ConfigurationBuilder().AddInMemoryCollection(new[]
+            {
+                new KeyValuePair<string, string>("Certificates:Development:Password", "1234"),
+            }).Build();
+
+            serverOptions.Configure(config);
+
+            Assert.Null(serverOptions.DefaultCertificate);
+
+            serverOptions.ConfigurationLoader.Load();
+
+            Assert.NotNull(serverOptions.DefaultCertificate);
+            Assert.Equal(serverOptions.DefaultCertificate.SerialNumber, certificate.SerialNumber);
+
+            var ran1 = false;
+            serverOptions.ListenAnyIP(4545, listenOptions =>
+            {
+                ran1 = true;
+                listenOptions.Protocols = HttpProtocols.Http1AndHttp2;
+                listenOptions.UseHttps();
+            });
+            Assert.True(ran1);
+
+            Assert.NotEmpty(serverOptions.CodeBackedListenOptions);
+            Assert.All(serverOptions.CodeBackedListenOptions, listenOptions =>
+            {
+                Assert.False(listenOptions.HttpsOptions.IsValueCreated);
+                listenOptions.Build();
+                Assert.True(listenOptions.HttpsOptions.IsValueCreated);
+                Assert.Equal(listenOptions.HttpsOptions.Value.ServerCertificate?.SerialNumber, certificate.SerialNumber);
+            });
+        }
+        finally
+        {
+            if (File.Exists(GetCertificatePath()))
+            {
+                File.Delete(GetCertificatePath());
+            }
+        }
+    }
+
+    [Fact]
+    public void LoadDevelopmentCertificate_UseHttpsFirst()
+    {
+        try
+        {
+            var serverOptions = CreateServerOptions();
+            var certificate = new X509Certificate2(TestResources.GetCertPath("aspnetdevcert.pfx"), "testPassword", X509KeyStorageFlags.Exportable);
+            var bytes = certificate.Export(X509ContentType.Pkcs12, "1234");
+            var path = GetCertificatePath();
+            Directory.CreateDirectory(Path.GetDirectoryName(path));
+            File.WriteAllBytes(path, bytes);
+
+            var ran1 = false;
+            serverOptions.ListenAnyIP(4545, listenOptions =>
+            {
+                ran1 = true;
+                listenOptions.Protocols = HttpProtocols.Http1AndHttp2;
+                listenOptions.UseHttps();
+            });
+            Assert.True(ran1);
+
+            var config = new ConfigurationBuilder().AddInMemoryCollection(new[]
+            {
+                new KeyValuePair<string, string>("Certificates:Development:Password", "1234"),
```

Comment:
Since HTTP/3 was a little tricky to make work, can we add end-to-end tests for this scenario to [Http3TlsTests](https://github.com/dotnet/aspnetcore/blob/5a4a8e4b456ad82ab4c3bdb798b3a5f4e2fb9706/src/Servers/Kestrel/test/Interop.FunctionalTests/Http3/Http3TlsTests.cs) where we verify that we can successfully bind and make requests in this scenario that does the docker-style dev cert loading instead of default cert loading. The Http3TlsTests require msquic support, so it might be easier to add a variant of [Http3_UseHttpsNoArgsWithDefaultCertificate_UseDefaultCertificate](
https://github.com/dotnet/aspnetcore/blob/5a4a8e4b456ad82ab4c3bdb798b3a5f4e2fb9706/src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsTests.cs#L409) instead. It should still be a little more end-to-end than these tests, but it uses a mock multiplexed transport and verifies the options flow correctly trhough BindAsync. I think adding a test to just one of these two test classes is fine. Do you have a preference @JamesNK?

For completeness, we could add a similar test [HttpsConnectionMiddlewareTests](https://github.com/dotnet/aspnetcore/blob/5a4a8e4b456ad82ab4c3bdb798b3a5f4e2fb9706/src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsConnectionMiddlewareTests.cs#L60) to test the end-to-end given a non-multiplexed in-memory transport.

We should also assert that we're still actually running the callback in all our UseHttps() tests that assert inside the callback. I suspect there are a number of tests where this is no longer happening. For instance, there's `ConfigureHttpsDefaultsNeverLoadsDefaultCert`,

https://github.com/dotnet/aspnetcore/blob/5a4a8e4b456ad82ab4c3bdb798b3a5f4e2fb9706/src/Servers/Kestrel/test/InMemory.FunctionalTests/HttpsTests.cs#L109-L119

Reply:
Did this part:
> We should also assert that we're still actually running the callback in all our UseHttps() tests that assert inside the callback

Reply:
Think I did this part:
> add end-to-end tests for this scenario to [Http3TlsTests](https://github.com/dotnet/aspnetcore/blob/5a4a8e4b456ad82ab4c3bdb798b3a5f4e2fb9706/src/Servers/Kestrel/test/Interop.FunctionalTests/Http3/Http3TlsTests.cs)

Reply:
I did this as practice, but it's just a simpler version of the above, so I reverted it:
> add a variant of Http3_UseHttpsNoArgsWithDefaultCertificate_UseDefaultCertificate

Reply:
Per our offline discussion, the current coverage seems adequate.